src/stream/src/executor/source/kafka_backfill_state_table.rs

```diff
@@ -0,0 +1,140 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Bound;
+
+use futures::{pin_mut, StreamExt};
+use risingwave_common::row;
+use risingwave_common::row::{OwnedRow, Row};
+use risingwave_common::types::{JsonbVal, ScalarImpl, ScalarRef, ScalarRefImpl};
+use risingwave_common::util::epoch::EpochPair;
+use risingwave_connector::source::SplitId;
+use risingwave_pb::catalog::PbTable;
+use risingwave_storage::StateStore;
+
+use super::kafka_backfill_executor::{BackfillState, BackfillStates};
+use crate::common::table::state_table::StateTable;
+use crate::executor::error::StreamExecutorError;
+use crate::executor::StreamExecutorResult;
+
+pub struct BackfillStateTableHandler<S: StateStore> {
+    pub state_store: StateTable<S>,
+}
+
+impl<S: StateStore> BackfillStateTableHandler<S> {
+    pub async fn from_table_catalog(table_catalog: &PbTable, store: S) -> Self {
+        Self {
+            state_store: StateTable::from_table_catalog(table_catalog, store, None).await,
+        }
+    }
+
+    pub fn init_epoch(&mut self, epoch: EpochPair) {
+        self.state_store.init_epoch(epoch);
+    }
+
+    fn string_to_scalar(rhs: impl Into<String>) -> ScalarImpl {
+        ScalarImpl::Utf8(rhs.into().into_boxed_str())
+    }
+
+    pub(crate) async fn get(&self, key: &SplitId) -> StreamExecutorResult<Option<OwnedRow>> {
+        self.state_store
+            .get_row(row::once(Some(Self::string_to_scalar(key.as_ref()))))
+            .await
+            .map_err(StreamExecutorError::from)
+    }
+
+    /// XXX: we might get stale data for other actors' writes, but it's fine?
+    pub async fn scan(&self) -> StreamExecutorResult<Vec<BackfillState>> {
+        let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) = &(Bound::Unbounded, Bound::Unbounded);
+
+        let state_table_iter = self
+            .state_store
+            .iter_with_prefix(None::<OwnedRow>, sub_range, Default::default())
+            .await?;
+        pin_mut!(state_table_iter);
+
+        let mut ret = vec![];
+        while let Some(item) = state_table_iter.next().await {
+            let row = item?.into_owned_row();
+            tracing::debug!("scanning backfill state table, row: {:?}", row);
+            let state = match row.datum_at(1) {
+                Some(ScalarRefImpl::Jsonb(jsonb_ref)) => {
+                    BackfillState::restore_from_json(jsonb_ref.to_owned_scalar())?
+                }
+                _ => unreachable!(),
+            };
+            ret.push(state);
+        }
+        Ok(ret)
+    }
+
+    pub async fn set(&mut self, key: SplitId, value: JsonbVal) -> StreamExecutorResult<()> {
```

Comment:
```suggestion
    pub async fn set(&mut self, key: SplitId, state: BackfillState) -> StreamExecutorResult<()> {
```

Should be strong-typed according to other methods. But this functions seems unused